pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,198 @@
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+const (
+	controllerName = "route_metrics_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route metrics controller. This is the controller
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager) (controller.Controller, error) {
+	// Create a copy of the cluster, but with a new cache to watch on Route objects from every namespace.
+	newCluster, err := cluster.New(mgr.GetConfig(), func(o *cluster.Options) {
+		o.Scheme = mgr.GetScheme()
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cluster to the manager so that the cluster is started along with the other runnables.
+	mgr.Add(newCluster)
+	reconciler := &reconciler{
+		client: mgr.GetClient(),
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCluster.GetCache()),
+		handler.EnqueueRequestsFromMapFunc(func(client.Object) []reconcile.Request { return nil }),
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return reconciler.routeCreateEventGatherMetrics(e.Object) },
+			DeleteFunc: func(e event.DeleteEvent) bool { return reconciler.routeDeleteEventGatherMetrics(e.Object) },
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				return reconciler.routeUpdateEventGatherMetrics(e.ObjectOld, e.ObjectNew)
+			},
+			GenericFunc: func(e event.GenericEvent) bool { return false },
+		}); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// routeCreateEventGatherMetrics gather RouteMetricsControllerRouteType and RouteMetricsControllerRoutesPerShard metrics
+// based on the Route Create event.
+func (r *reconciler) routeCreateEventGatherMetrics(obj client.Object) bool {
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	log.Info("route create event", "name", routeObject.Name, "namespace", routeObject.Namespace)
+
+	// Iterate through the related Ingress Controllers.
+	for _, ingress := range routeObject.Status.Ingress {
```

Comment:
The canary controller has a [function](https://github.com/openshift/cluster-ingress-operator/blob/542124811a01039852d8b0846da91d33665037f8/pkg/operator/controller/canary/route.go#L163) that does this check but only for the default IC. It might be worth reusing that strategy and abstracting the logic to it's own function. I haven't tested this code but it'd look something like this:

```go
// routeAdmitted returns true if a given route has been admitted.
func routeAdmitted(route *routev1.Route) bool {
	for _, ingress := range route.Status.Ingress {
		for _, cond := range ingress.Conditions {
			if cond.Type == routev1.RouteAdmitted && cond.Status == corev1.ConditionTrue {
				return true
			}
		}
	}
	return false
}
```
Then `routeCreateEventGatherMetrics` would look like this:
```go
if routeAdmitted(routeObject) {
  IncrementRouteMetricsControllerRoutesPerShardMetric(ingress.RouterName)
}
```
The main advantage being that you get to remove several lines from the `route*EventGatherMetrics` functions.

Reply:
Thanks for the suggestion. I have added the `routeAdmitted` function.